// @protocolink/logics/dist/logics/radiant-v2/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractAddress = exports.configMap = exports.supportedChainIds = exports.configs = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.configs = [
    {
        chainId: common.ChainId.mainnet,
        contract: {
            ProtocolDataProvider: '0x362f3BB63Cff83bd169aE1793979E9e537993813',
            RadiantV2FlashLoanCallback: '0x6bfCE075A1c4F0fD4067A401dA8f159354e1a916',
        },
    },
    {
        chainId: common.ChainId.bnb,
        contract: {
            ProtocolDataProvider: '0x2f9D57E97C3DFED8676e605BC504a48E0c5917E9',
            RadiantV2FlashLoanCallback: '0x6bfCE075A1c4F0fD4067A401dA8f159354e1a916',
        },
    },
    {
        chainId: common.ChainId.arbitrum,
        contract: {
            ProtocolDataProvider: '0x596B0cc4c5094507C50b579a662FE7e7b094A2cC',
            RadiantV2FlashLoanCallback: '0x6bfCE075A1c4F0fD4067A401dA8f159354e1a916',
        },
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getContractAddress(chainId, name) {
    return exports.configMap[chainId].contract[name];
}
exports.getContractAddress = getContractAddress;
